clients/drcachesim/tests/burst_longblock.cpp

```diff
@@ -0,0 +1,323 @@
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t" // 180
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t" // 190
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t"
+                             "fpatan\n\t" // 200
+                             "mov (%rsp),%rax\n\t");
+#else
+        // It is not worth the effort to replicate on other platforms.
+#    error Not implemented
+#endif
+    }
+    return nullptr;
+}
+
+static int
+do_some_work()
+{
+    // To reduce the trace size we just sleep and let the other thread
+    // be the only one generating data.
+    constexpr int MILLIS_TO_SLEEP = 10;
+    std::this_thread::sleep_for(std::chrono::milliseconds(MILLIS_TO_SLEEP));
+    return 1;
+}
+
+static void
+gather_trace()
+{
+    // We need -no_align_endpoints to make reproducing the bug much more likely.
+    // Otherwise, the tracer switches to nop mode and the detach happens mid-block
+    // but tracing has already ended.
+    if (!my_setenv("DYNAMORIO_OPTIONS",
+                   "-stderr_mask 0xc -client_lib ';;-offline -no_align_endpoints"))
+        std::cerr << "failed to set env var!\n";
+    std::cerr << "pre-DR init\n" << std::flush;
+    dr_app_setup_and_start();
+    assert(dr_app_running_under_dynamorio());
+    if (do_some_work() < 0)
+        std::cerr << "error in computation\n";
+    // TODO i#6490: This app produces incorrect output when run under DR if we do
+    // not flush. std::endl makes the issue worse even though it should do an
+    // internal flush.
+    std::cerr << "pre-DR detach\n" << std::flush;
+    dr_app_stop_and_cleanup();
+    std::cerr << "all done\n" << std::flush;
+}
+
+int
+test_main(int argc, const char *argv[])
+{
+    // Start up a thread that spends most of its time in a long block.
+    thread_t thread = create_thread(thread_func, nullptr);
+    // Now gather a trace where the detach should often split the block.
+    gather_trace();
+    child_should_exit.store(true, std::memory_order_release);
+    join_thread(thread);
+    // The test harness will post-process the trace and run invariant_checker
```

Comment:
So ideally cmake should run this test Nx and ensure there isn't a single failure? (Assuming the odds of hitting the mid-block-detach are very high 1/N).

Reply:
Actually this test hits the problem every single time: 50 out of 50 when I tested it. So one run is enough. It's because fpatan is a very slow instruction :)

I added here:

    // As noted, the test hits a mid-block detach reliably every time,
    // so we just need the one trace.

I added up above:

    // (it happens in 50 out of 50 runs this way and almost never with
    // -align_endpoints enabled).